MWI: Add lock targets for join token name and bot instance ID

timothyb89 opened the pull request:
This adds two new lock targets meant to help lock specific bot instances without affecting all bots sharing a single user:
- Bot Instance ID: Targets a bot instance UUID, which has been assigned automatically to unique bot instances for some time
- Join token name: Targets the join token through which the bot joined

Bot instance ID locks are most useful for traditional token-joined bots, since tokens are single use and bots have no way to onboard again without human intervention if their old certs (and old bot instance) expire.

Join token locks are useful for bots using delegated join methods. They are particularly useful for bound keypair joining, where there is a direct 1:1 relationship between a "bot instance" and a token, even though that bot ID will change each time a recovery takes place.

Note that this does not currently set the join token for nodes even though that would theoretically be possible. We could consider supporting node locking in the future if there's demand.

timothyb89 committed:
MWI: Add lock targets for join token name and bot instance ID

This adds two new lock targets meant to help lock specific bot
instances without affecting all bots sharing a single user:
- Bot Instance ID: Targets a bot instance UUID, which has been
  assigned automatically to unique bot instances for some time
- Join token name: Targets the join token through which the bot
  joined

Bot instance ID locks are most useful for traditional token-joined
bots, since tokens are single use and bots have no way to onboard
again without human intervention if their old certs (and old bot
instance) expire.

Join token locks are useful for bots using delegated join methods.
They are particularly useful for bound keypair joining, where there
is a direct 1:1 relationship between a "bot instance" and a token,
even though that bot ID will change each time a recovery takes place.

Note that this does not currently set the join token for nodes even
though that would theoretically be possible. We could consider
supporting node locking in the future if there's demand.

timothyb89 committed:
Set join token cert request field for non-renewable bot identities

timothyb89 committed:
Fix ASN ID and pass through join token name in impersonated certs

timothyb89 committed:
Tweak docstrings and add missing references for lib/decision

juliaogris commented on api/proto/teleport/decision/v1alpha1/ssh_access.proto:
```diff
@@ -159,6 +159,14 @@ message LockTarget {
+
+  // BotInstanceID is the bot instance ID if this is a bot identity; it is
+  // ignored otherwise.
+  string bot_instance_id = 9;
```
If `bot_instance_id` and `join_token` can only exist together, should they be combined in a separate message?
What happens to the `user` field in case of a bot instance - is this worth a comment?

timothyb89 replied:
It's a bit messy but it probably doesn't make sense to move these fields into a separate message. 

Bots are a variant of users and have a username (`bot-foo`, generally), so `user` will be set for both human users and bots. All bots have a `bot_instance_id`, but only some bots will have a `join_token` set. The fields will be empty for regular users, which is a convention shared with other fields, e.g. `server_id` is only set for servers, `mfa_device` is only set when a user has an MFA device, etc.

timothyb89 committed:
Clarify docstrings

Clarifies various docstrings and makes sure they mention `token`
joined bots cannot be targeted.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bound-keypair-lock-targets

timothyb89 committed:
Fix failing tests
